docs/architecture/adr-004-tips-contract.md

```diff
@@ -0,0 +1,141 @@
+# ADR 004: Tips Contract
+
+## Changelog
+
+- Aug 25, 2022: Initial draft.
+
+## Status
+DRAFTED
+
+## Abstract
+This ADR defines the architecture of the Tips contract. This contract manage sending and tracking of tips made within
+Desmos based applications and exposes queries that allow to gather all users sent and received tips easily.
+
+## Context
+Desmos based applications may want to leverage on the web3 features and one of the most popular one is for sure the possibility to tip users. Currently, it's easy to integrate a feature that allow an application to send tokens as tips between users but very difficult to track those tips in an easy way.
+Plus application creators might want to integrate some form of monetization that allows them to earn a little percentage when user send tips to each other.
+
+## Decision
+The idea here is to write a contract called `Tips`, that handles the sending and tracking of user's tips.
+This contract allows sending tips between users as well tracking sent and received tips for each one of them.
+The tips tracking should be limited to a maximum tracked record number that helps to keep the contract storage as small
+as possible. Also, it doesn't make much sense to store all the tips record that a user received because each one of them will
+most likely check only the latest received.
+Additionally, the contract allows its admin to set a fee that users has to pay in order to be allowed to send a tip. This fee, if not set, is equal to zero.
+
+## Specifications
+Here below the specifications for the contract's messages:
+
+### Messages
+
+#### Instantiate
+```rust
+pub struct InstantiateMsg {
+  pub admin: String,
+  pub subspace_id: u64,
+  pub service_fee: Vec<Coin>,
```

Comment:
Currently the service fees are set as a list of `Coin`. However, there might be cases in which a percentage fee would be better. Should we allow setting both kind of fees (percentage, and fixed)?

Reply:
Percentage means a % based on the tips the user give? 
If so, I think it can be done without problems

Reply:
@bragaz Yes. There might be cases in which contract owners want to get a cut on the tips in percentage rather thank fixed value. I think allowing them to decide what kind of fee type to use (percentage or fixed) would be better. In this way we can support multiple use cases.

_Fixed_  
The contract owner sets a fixed amount of service fees to be paid (e.g `0.01` DSM) each time a user wants to send a tip.

_Percentage_ 
The contract owner sets a percentage amount of service fees to be paid (e.g. `10%`) each time a user wants to send a tip. The service fee amount will then be computed as `service_fee_amount = service_fee_percentage * tip_amoun`